connect/mirror/src/main/java/org/apache/kafka/connect/mirror/rest/resources/InternalConnectorsResource.java

```diff
@@ -0,0 +1,120 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.kafka.connect.mirror.rest.resources;
+
+import com.fasterxml.jackson.core.type.TypeReference;
+import com.fasterxml.jackson.databind.ObjectMapper;
+import io.swagger.v3.oas.annotations.Operation;
+import org.apache.kafka.common.config.AbstractConfig;
+import org.apache.kafka.connect.mirror.SourceAndTarget;
+import org.apache.kafka.connect.runtime.Herder;
+import org.apache.kafka.connect.runtime.rest.HerderRequestHandler;
+import org.apache.kafka.connect.runtime.rest.InternalRequestSignature;
+import org.apache.kafka.connect.runtime.rest.resources.ConnectResource;
+import org.apache.kafka.connect.util.FutureCallback;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import javax.ws.rs.NotFoundException;
+import javax.ws.rs.POST;
+import javax.ws.rs.PUT;
+import javax.ws.rs.Path;
+import javax.ws.rs.PathParam;
+import javax.ws.rs.Produces;
+import javax.ws.rs.QueryParam;
+import javax.ws.rs.core.Context;
+import javax.ws.rs.core.HttpHeaders;
+import javax.ws.rs.core.MediaType;
+import java.util.List;
+import java.util.Map;
+
+@Path("/")
+@Produces(MediaType.APPLICATION_JSON)
+public class InternalConnectorsResource implements ConnectResource {
+
+    private static final Logger log = LoggerFactory.getLogger(InternalConnectorsResource.class);
+
+    private static final TypeReference<List<Map<String, String>>> TASK_CONFIGS_TYPE =
+            new TypeReference<List<Map<String, String>>>() { };
+
+    private final Map<SourceAndTarget, Herder> herders;
+    private final HerderRequestHandler requestHandler;
+
+    public InternalConnectorsResource(Map<SourceAndTarget, Herder> herders, AbstractConfig serverConfig) {
+        this.herders = herders;
+        this.requestHandler = new HerderRequestHandler(serverConfig, DEFAULT_REST_REQUEST_TIMEOUT_MS);
+    }
+
+    @Override
+    public void requestTimeout(long requestTimeoutMs) {
+        requestHandler.requestTimeoutMs(requestTimeoutMs);
+    }
+
+    @POST
+    @Path("/{source}/{target}/connectors/{connector}/tasks")
```

Comment:
Rather than re-implementing these paths/signatures/method calls, can we re-use them with JAX-RS sub-resources and sub-resource-loaders?

I imagine that we'd have a non-root (without `@Path("/")`) resource class with implementations of these methods (with `@Path("/connectors/{connector}/tasks")` for example.) An object of that class would be returned by a sub-resource-loader method tagged with `@Path("/{source}/{target}")` from this class, and would also be subclassed in a root-resource (with `@Path("/")`) to be loaded into the normal distributed-mode server.

That way there's only one `putTaskConfigs` method, but it's implementation is accessible from `/` in distributed mode and `/source/target/` in MM mode.

Reply:
It's not quite that simple, since the implementations of `fenceZombies` and `putTaskConfigs` actually differ slightly across the two resources. The `ConnectorsResource` implementation uses a single herder, whereas the `InternalConnectorsResource` implementation has multiple herders and chooses a single one based on the `{source}` and `{target}` in the URLs for the paths.

There's still a chance to pull some of the shared logic out into a single class, but at that point I'm not sure how much effort we're really saving.

Reply:
Never mind, this totally works. I forgot that you can fetch path information from an injected `UriInfo` field instead of having to grab it from `@PathParam`-annotated method parameters. I've implemented this in the latest commit; LMKWYT.